### ISSUE TITLE
storage: (*Replica).handleEvalResult called on Replica with zero in-memory, nonzero on-disk Lease

### ISSUE THREAD

**Issue by tschottdorf** [cockroachdb/cockroach#15819]
We were calling `handleEvalResult` on a replica that had zero Lease state, but nonzero on-disk Lease state.

```
F170509 20:02:12.165939 136 storage/replica.go:1353  [n5,s5,r2/5:/System/{-tsd}] on-disk and in-memory state diverged:
[Lease.Start.WallTime: 1494353965502260882 != 0 Lease.Expiration.WallTime: 1494360140105540031 != 0 Lease.Replica.NodeID: 3 != 0 Lease.Replica.StoreID: 3 != 0 Lease.Replica.ReplicaID: 3 != 0 Lease.DeprecatedStartStasis.WallTime: 1494360140105540031 != 0 Lease.ProposedTS: &hlc.Timestamp{WallTime:1494360131105562632, Logical:0} != nil]
```

Stack trace:

```
goroutine 136 [running]:
github.com/cockroachdb/cockroach/pkg/util/log.getStacks(0x1, 0x9e40b1c, 0x2c18620, 0x10)
        /go/src/github.com/cockroachdb/cockroach/pkg/util/log/clog.go:858 +0xa7
github.com/cockroachdb/cockroach/pkg/util/log.(*loggingT).outputLogEntry(0x2c19e00, 0xc400000004, 0x24e206d, 0x12, 0x549, 0xc43631e000, 0x191)
        /go/src/github.com/cockroachdb/cockroach/pkg/util/log/clog.go:726 +0x56d
github.com/cockroachdb/cockroach/pkg/util/log.addStructured(0x7f19ea15bd38, 0xc420cc2a50, 0x7f1900000004, 0x2, 0x1b4c4ef, 0x28, 0xc432ebb8a0, 0x1, 0x1)
        /go/src/github.com/cockroachdb/cockroach/pkg/util/log/structured.go:158 +0x4b3
github.com/cockroachdb/cockroach/pkg/util/log.logDepth(0x7f19ea15bd38, 0xc420cc2a50, 0x1, 0xc400000004, 0x1b4c4ef, 0x28, 0xc432ebb8a0, 0x1, 0x1)
        /go/src/github.com/cockroachdb/cockroach/pkg/util/log/log.go:68 +0x8c
github.com/cockroachdb/cockroach/pkg/util/log.Fatalf(0x7f19ea15bd38, 0xc420cc2a50, 0x1b4c4ef, 0x28, 0xc432ebb8a0, 0x1, 0x1)
        /go/src/github.com/cockroachdb/cockroach/pkg/util/log/log.go:162 +0x7e
github.com/cockroachdb/cockroach/pkg/storage.(*Replica).assertStateRLocked(0xc420cd6000, 0x27e7060, 0xc4203f60b0)
        /go/src/github.com/cockroachdb/cockroach/pkg/storage/replica.go:1353 +0x40a
github.com/cockroachdb/cockroach/pkg/storage.(*Replica).assertState(0xc420cd6000, 0x27e7060, 0xc4203f60b0)
        /go/src/github.com/cockroachdb/cockroach/pkg/storage/replica.go:1338 +0x8c
github.com/cockroachdb/cockroach/pkg/storage.(*Replica).handleEvalResult(0xc420cd6000, 0x7f19ea15bd38, 0xc420cc2a50, 0x0, 0xc42d6ca610, 0xe, 0x10, 0xc42d6ca620, 0xf, 0x10, ...)
        /go/src/github.com/cockroachdb/cockroach/pkg/storage/replica_proposal.go:790 +0x113
github.com/cockroachdb/cockroach/pkg/storage.(*Replica).processRaftCommand(0xc420cd6000, 0x7f19ea15bd38, 0xc420cc2a50, 0xc42d6ca5f0, 0x8, 0x29ba8b, 0x300000003, 0x3, 0x14bd0303a2756692, 0x0, ...)
        /go/src/github.com/cockroachdb/cockroach/pkg/storage/replica.go:3711 +0x837
github.com/cockroachdb/cockroach/pkg/storage.(*Replica).handleRaftReadyRaftMuLocked(0xc420cd6000, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
        /go/src/github.com/cockroachdb/cockroach/pkg/storage/replica.go:2878 +0xd70
github.com/cockroachdb/cockroach/pkg/storage.(*Store).processRaftRequest(0xc420cbc000, 0x7f19ea15bd38, 0xc4330c6390, 0xc438324480, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
        /go/src/github.com/cockroachdb/cockroach/pkg/storage/store.go:3110 +0xcc5
github.com/cockroachdb/cockroach/pkg/storage.(*Store).processRequestQueue(0xc420cbc000, 0x2)
        /go/src/github.com/cockroachdb/cockroach/pkg/storage/store.go:3389 +0x172
github.com/cockroachdb/cockroach/pkg/storage.(*raftScheduler).worker(0xc420183ef0, 0xc4206606e0)
        /go/src/github.com/cockroachdb/cockroach/pkg/storage/scheduler.go:228 +0x24b
github.com/cockroachdb/cockroach/pkg/storage.(*raftScheduler).Start.func2(0x7f19ea15bd38, 0xc421c36300)
        /go/src/github.com/cockroachdb/cockroach/pkg/storage/scheduler.go:168 +0x33
github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).RunWorker.func1(0xc4216984d0, 0xc4206606e0, 0xc421c29340)
        /go/src/github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:215 +0xf7
created by github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).RunWorker
        /go/src/github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:216 +0xad
```

**Comment by tamird** [comment]
Where did this happen? How did you discover it?

**Comment by bdarnell** [comment]
It happened on cyan just after being upgrade from 725e5b3 to 220feef9 ([log](https://github.com/cockroachdb/cockroach/compare/752e5b3...220feef9)). #15783 seems like the most likely culprit.

**Comment by petermattis** [comment]
> How did you discover it?

Panic/log-fatal reporting via sentry.io!

**Comment by tschottdorf** [comment]
I browsed #15783 but couldn't find anything obvious. @tamird, could you sleuthe (is that a word?
 should be anyway) this? cyan-5 was running 752e5b3, and supposedly talked to something running 220feef (i.e. non-nullable `Lease`).

**Comment by tschottdorf** [comment]
(in the meantime, should we revert #15783? It looks like a relatively likely culprit, except if there's something older in the commit log that just hadn't fired until now. Perhaps downgrading a node in cyan triggers this again right away?)

**Comment by tamird** [comment]
We can revert https://github.com/cockroachdb/cockroach/commit/c29363039ba7d116dcb65f8dd379058fe961fefb (no need to revert the other commits in that PR).

I think I initially misread this issue - it's somewhat worrying that this replica had a zero in-memory state and attempted to to `assertStateRLocked`. I'm looking.

**Comment by tschottdorf** [comment]
Confusion was on my part (I was also confused) - the assertion fires because the replica has empty in-memory lease, but not empty on disk.

**Comment by tschottdorf** [comment]
@tamird and I figured it out, he's taking it from here. TL;DR: revert the nullability, and future-proof so we can change it in the distant future.

**Comment by tamird** [comment]
Thanks for the help, @tschottdorf!

**Comment by andreimatei** [comment]
I'm glad y'all figured this out!

I know it's easy to kibitz (that word!) from the sidelines, but I think we should take the opportunity to:
a) invent a method of tagging issues or maintain a list of cleanup like this field nullability change that we can perform the next time we have to break compatibility for a different reason. What do you think?
b) have a mixed-version test that would have caught this. Apparently the only mixed-version test we have is about writing data with one version and reading it with another. Here we need one where raft proposals from one version (in particular, the new one) are processed by the other. This has always seemed like a big testing gap to me. cc @petermattis

**Comment by petermattis** [comment]
> have a mixed-version test that would have caught this.

Can you file an issue about this? There might already be such an issue. If there is, ping it as the priority is higher now.

**Comment by bdarnell** [comment]
We don't yet have a mixed-version test that runs two versions at the same time. We need one, definitely before 1.1 and probably early in that cycle. 

> maintain a list of cleanup like this field nullability change that we can perform the next time we have to break compatibility for a different reason

I'm skeptical that such an issues list will ever be useful. We can't break compatibility any more; all changes will need to include their own migrations for a smooth transition. If it's not worth fixing now and implementing that migration, it's unlikely that that will ever get easier or more worthwhile in the future, so these issues will just clutter up the tracker. If it so happens that we change the lease implementation that fixing this issue is a natural side effect of other changes, then we'll clean this up whether there's an issue or not. But if it's not so easy, I don't think we will or should do it.

**Comment by andreimatei** [comment]
Ben, we can't break compatibility from version to the next, but that doesn't mean that we can't break compatibility between any versions, does it? In this case, if we modify the check, as Tamir did, to catch both nil and empty leases, then we'll be able to change the field when compatibility with the current version is being broken.
Or, I guess, if you don't agree and insist on a stricter backwards compatibility story, then we shouldn't change the check in this way, but I understand that this kind of preparation for the future has been a practice before.

**Comment by bdarnell** [comment]
Yeah, we can eventually change things over multiple releases (years!). So we can start the process now and lay the groundwork for a future change. But an issue on its own without a current change to start the process doesn't make much sense. 

And I'd say that in general, changes like this are not worth the trouble of long-term coordination. In most cases it's better to live with a suboptimal design. There will always be more important things to work on and each half-finished transition in the code has a cost. The future benefit will need to be substantial.

**Comment by andreimatei** [comment]
Peter, I couldn't find an issue about mixed-version testing. Filed #15851.

**Comment by jordanlewis** [comment]
https://sentry.io/cockroach-labs/cockroachdb/issues/270694056/

**Comment by tamird** [comment]
@jordanlewis that occurred on cyan-3. Here's the story:
- c293630 introduced a backwards incompatibility that would cause a cluster to crash while it was being upgraded, but once all nodes were upgraded, things worked fine.
- d0bd273 reverted that backwards incompatibility, but did it in such a way that was itself backwards incompatible with c293630.

Cyan had already been "fully upgraded" to c293630, and so could not be upgraded to d0bd273 without being wiped. It is now being wiped.

TL;DR this should be resolved now.
